src/runtime_src/xocl/api/clReleaseCommandQueue.cpp

```diff
@@ -37,6 +37,13 @@ static cl_int
 clReleaseCommandQueue(cl_command_queue command_queue)
 {
   validOrError(command_queue);
+  // validOrError checks for nullptr only in the debug build. Suppress
+  // collateral GCC warnings (e.g. stringop-overflow in smart pointer methods)
+  // in case of a non-debug build.
+  #ifdef NDEBUG
+  if (!command_queue) __builtin_unreachable();
+  #endif
```

Comment:
I don't really like this conditional code.  Prefer to disable the stringop-overflow in the file itself.   It's just not very useful code checking for a condition that has already been checked even in optimized compile, .e.g. validOrError is always enabled, and users have to jump through hoops to disable it.

Reply:
>  Prefer to disable the stringop-overflow in the file itself

If disabled, then gcc cannot check for stringop-overflow in debug (i.e., non-optimized) build.

>  It's just not very useful code checking for a condition that has already been checked even in optimized compile, .e.g. validOrError is always enabled, and users have to jump through hoops to disable it.

As I mentioned in [my previous comment](https://github.com/Xilinx/XRT/pull/6984#issuecomment-1235504053), nullptr check (probably `validOrError`) does not seem to be active in the release (i.e., optimized) build. Please correct me if I am wrong.

So I still don't see the advantage of disabling the warning on the file level. The behavior of the code seem to differ between release and debug builds.

Reply:
Thanks @goekce for pinging me.

Definitely validOrError is active in release builds. There is config run-time option to disable validOrError.  I do not understand why gcc would complain?  IMO it smells like an compiler optimization bug (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=104475). 

I am able to reproduce the gcc warning with a small stand-alone program, but it really just confirms that optimization is wrong.

I understand the work-around (it is mentioned in the gcc bug report also), but it will not work as is since XRT is not for GCC only.   Also we would have to fix the comment given that validOrError does indeed check for nullptr both in DEBUG and NDEBUG, but we are facing a compiler bug.

Let's keep the conditional macro code in one file.  I suggest adding `xocl_or_error()` in `src/runtime_src/xocl/core/object.h` and replace the `xocl(...)` calls in the `ZZZRelease.cpp` files with `xocl_or_error(...)` when we want to enforce undefined behavior when the cl object is null.

```
template <typename CLTYPE>
typename detail::cl_object_traits<CLTYPE>::xocl_type*
xocl_or_error(CLTYPE c)
{
  auto x = detail::cl_object_traits<CLTYPE>::get_xocl(c);
#ifdef __GNUC__
  if (!x)
    __builtin_unreachable();
#endif
  return x;
}

```